docs/src/main/asciidoc/cdi.adoc

```diff
@@ -0,0 +1,361 @@
+A client proxy is basically an object that delegates all method invocations to a target bean instance.
+It's a container construct that implements `io.quarkus.arc.ClientProxy` and extends the bean class.
+The client proxy instance is always injected instead of a contextual instance of a normal scoped bean (e.g. `@ApplicationScoped` and `@RequestScoped`).
+Client proxies allow for:
+
+* Lazy instantiation - the instance is created once a method is invoked upon the proxy.
+* Ability to inject a bean with "narrower" scope to a bean with "wider" scope; i.e. you can inject a `@RequestScoped` bean into an `@ApplicationScoped` bean.
+* Circular dependencies in the dependency graph. Circular dependencies are "bad design" but sometimes they're inevitable. 
+* In rare cases it's practical to destroy the beans manually. A direct injected reference would lead to a stale bean instance.
+ 
+ 
+== _Q: OK. You said that there are several kinds of beans?_
+
+A: Yes. In general, in Quarkus we distinguish:
+
+1. Class beans
+2. Producer methods
+3. Producer fields
+4. Synthetic beans
+
+Producer methods are useful if you need additional control over instantiation of a bean.
+They are also useful when integrating third-party libraries where you don't control the class source and may not add additional annotations etc.
+
+.Producer Method Example
+[source,java]
+----
+@ApplicationScoped
+public class PiProducer {
+
+    @Produces <1>
+    double producePi() { 
+       return Math.PI; <2>
+    }
+}
+
+@ApplicationScoped
+public class PiConsumer {
+
+   @Inject
+   double pi; <3>
+   
+   double twoPis() {
+     return pi * 2;
+   }
+} 
+----
+<1> The container analyses the method annotations to build a bean metadata.
+The return type is used to build the set of bean types. 
+In this case, it will be `double` and `java.lang.Object`.
+No scope annotation is declared and so it's defaulted to `@Dependent`.
+<2> The container will delegate to this method when creating the bean instance.
+<3> `PiProducer#producePi()` is invoked and the result is set to `PiConsumer#pi` when the `PiConsumer` instance is created.
+
+NOTE: One of the limitations of producers is that interceptors are not applied to the result of a producer method/field.
+
+Producer fields are very similar to producer methods except that the field value is read when obtaining a bean instance.
+
+TIP: There's more about producers. You can declare qualifiers, inject dependencies into the producer methods parameters, etc. You can read more about producers for example in the https://docs.jboss.org/weld/reference/latest/en-US/html/beanscdi.html#_producer_methods[Weld docs, window="_blank"].
+
+Synthetic beans are usually provided by extensions.
+A synthetic bean is not backed by a class/method/field.
+Application developers usually only consume this kind of beans.
```

Comment:
As it didn't target developers, maybe don't introduce them ? It's and introduction guide we don't need to cover everything

Reply:
Yeah, I think they shouldn't be introduced here.

Reply:
I am biased about it. On one hand developers only consume them, OTOH they need to know they exist otherwise they will keep seeing injection of beans that have no classes/producers backing them and are seemingly conjured out of thin air.

Reply:
+1

Reply:
I only added a note to the list...